File: services/analytics/app/domain/contracts.py
"""Mirrors packages/domain/src/contracts.ts. Both are validated against the
same fixtures in data/schemas/fixtures (see tests/test_contracts.py and
packages/domain/src/contracts.test.ts) so drift is caught by CI."""

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

from app.domain.enums import Direction


class LeagueScoringRule(BaseModel):
    stat: str = Field(min_length=1, max_length=64)
    weight: float
    direction: Direction
    punt: bool


class ProjectedLine(BaseModel):
    player_id: UUID = Field(alias="playerId")
    games: float = Field(ge=0)
    minutes_per_game: float = Field(ge=0, alias="minutesPerGame")
    pts: float = Field(ge=0)
    reb: float = Field(ge=0)
    ast: float = Field(ge=0)
    stl: float = Field(ge=0)
    blk: float = Field(ge=0)
    tov: float = Field(ge=0)
    fgm: float = Field(ge=0)
    fga: float = Field(ge=0)
    ftm: float = Field(ge=0)
    fta: float = Field(ge=0)
    three_pm: float = Field(ge=0, alias="threePm")
    lower80: dict[str, float]
    upper80: dict[str, float]
    injury_risk: float = Field(ge=0, le=1, alias="injuryRisk")
    consistency: float = Field(ge=0, le=1)
    upside: float = Field(ge=0, le=1)
    role_security: float = Field(ge=0, le=1, alias="roleSecurity")

    model_config = {"populate_by_name": True}


ScoreComponentKey = Literal[
    "production",
    "scarcity",
    "rosterNeed",
    "risk",
    "consistency",
    "age",
    "adpValue",
    "upside",
    "role",
    "nextPickAvailability",
    "preference",
    "schedule",
]


class ScoreComponent(BaseModel):
    key: ScoreComponentKey
    raw: float
    normalized: float = Field(ge=0, le=1)
    weight: float = Field(ge=0, le=1)
    contribution: float
    reason: str = Field(min_length=1)


RecommendationLabel = Literal[
    "BEST_OVERALL", "BEST_FIT", "BEST_VALUE", "HIGHEST_UPSIDE", "SAFEST_PICK"
]
Confidence = Literal["LOW", "MEDIUM", "HIGH"]


class Recommendation(BaseModel):
    player_id: UUID = Field(alias="playerId")
    rank: Literal[1, 2, 3]
    draft_score: float = Field(ge=0, le=100, alias="draftScore")
    labels: list[RecommendationLabel] = Field(min_length=1)
    explanation: str = Field(min_length=1)
    availability_next_pick: float = Field(ge=0, le=1, alias="availabilityNextPick")
    confidence: Confidence
    components: list[ScoreComponent] = Field(min_length=1)
    engine_version: str = Field(min_length=1, alias="engineVersion")
    projection_run_id: UUID = Field(alias="projectionRunId")
    input_checksum: str = Field(min_length=1, alias="inputChecksum")

    model_config = {"populate_by_name": True}


__all__: list[str] = [
    "LeagueScoringRule",
    "ProjectedLine",
    "ScoreComponent",
    "ScoreComponentKey",
    "Recommendation",
    "RecommendationLabel",
    "Confidence",
]

File: packages/domain/src/contracts.ts
import { z } from "zod";
import { directionSchema } from "./enums";

/**
 * Shared contracts exchanged between the TypeScript app layer and the Python
 * analytics service, per BUILD_SPEC.md section 5. Both sides validate the same
 * fixtures in data/schemas/fixtures — see packages/domain/src/contracts.test.ts
 * and services/analytics/tests/test_contracts.py.
 *
 * These are structural contracts only (Phase 0). The recommendation engine,
 * projection pipeline, and persistence that produce/consume real instances of
 * these shapes are built in later phases.
 */

const wellKnownStatKeys = [
  "PTS",
  "REB",
  "AST",
  "STL",
  "BLK",
  "TOV",
  "FG_PCT",
  "FT_PCT",
  "THREE_PM",
] as const;

/** League scoring category or points weight. `stat` accepts the well-known keys above
 * or a league-defined custom category key. */
export const leagueScoringRuleSchema = z.object({
  stat: z.string().min(1).max(64),
  weight: z.number(),
  direction: directionSchema,
  punt: z.boolean(),
});
export type LeagueScoringRule = z.infer<typeof leagueScoringRuleSchema>;
export { wellKnownStatKeys };

const statRecordSchema = z.record(z.string(), z.number());

export const projectedLineSchema = z.object({
  playerId: z.uuid(),
  games: z.number().nonnegative(),
  minutesPerGame: z.number().nonnegative(),
  pts: z.number().nonnegative(),
  reb: z.number().nonnegative(),
  ast: z.number().nonnegative(),
  stl: z.number().nonnegative(),
  blk: z.number().nonnegative(),
  tov: z.number().nonnegative(),
  fgm: z.number().nonnegative(),
  fga: z.number().nonnegative(),
  ftm: z.number().nonnegative(),
  fta: z.number().nonnegative(),
  threePm: z.number().nonnegative(),
  lower80: statRecordSchema,
  upper80: statRecordSchema,
  injuryRisk: z.number().min(0).max(1),
  consistency: z.number().min(0).max(1),
  upside: z.number().min(0).max(1),
  roleSecurity: z.number().min(0).max(1),
});
export type ProjectedLine = z.infer<typeof projectedLineSchema>;

export const scoreComponentKeySchema = z.enum([
  "production",
  "scarcity",
  "rosterNeed",
  "risk",
  "consistency",
  "age",
  "adpValue",
  "upside",
  "role",
  "nextPickAvailability",
  "preference",
  "schedule",
]);
export type ScoreComponentKey = z.infer<typeof scoreComponentKeySchema>;

export const scoreComponentSchema = z.object({
  key: scoreComponentKeySchema,
  raw: z.number(),
  normalized: z.number().min(0).max(1),
  weight: z.number().min(0).max(1),
  contribution: z.number(),
  reason: z.string().min(1),
});
export type ScoreComponent = z.infer<typeof scoreComponentSchema>;
// Structural Zod schema mirrors the engine's ScoreComponent interface
// (recommendation.ts) — the engine type remains the implementation source.

export const recommendationLabelSchema = z.enum([
  "BEST_OVERALL",
  "BEST_FIT",
  "BEST_VALUE",
  "HIGHEST_UPSIDE",
  "SAFEST_PICK",
]);
export type RecommendationLabel = z.infer<typeof recommendationLabelSchema>;

export const confidenceSchema = z.enum(["LOW", "MEDIUM", "HIGH"]);
export type Confidence = z.infer<typeof confidenceSchema>;

export const recommendationSchema = z.object({
  playerId: z.uuid(),
  rank: z.union([z.literal(1), z.literal(2), z.literal(3)]),
  draftScore: z.number().min(0).max(100),
  labels: z.array(recommendationLabelSchema).min(1),
  explanation: z.string().min(1),
  availabilityNextPick: z.number().min(0).max(1),
  confidence: confidenceSchema,
  components: z.array(scoreComponentSchema).min(1),
  engineVersion: z.string().min(1),
  projectionRunId: z.uuid(),
  inputChecksum: z.string().min(1),
});
export type Recommendation = z.infer<typeof recommendationSchema>;
